src/app/js/formats/vega/component/tree-map/TreeMapData.js

```diff
@@ -0,0 +1,166 @@
+    idIncrement;
+    /**
+     * List of nodes and leaves ids
+     * @type {Map<string, number>}
+     */
+    ids;
+    /**
+     * Raw list of nodes and leaves ready to be filtered and transformed
+     * @type {Map<string, {parent: number?, size: number?}>}
+     */
+    rawNodesAndLeaves;
+    /**
+     * Formatted list of nodes and leaves ready to be cleanup and returned
+     * @type {Map<number, {id: number, name: string, parent: number?, size: number?}>}
+     */
+    formattedNodesAndLeaves;
+    /**
+     * Filtered nodes use to create the final transformed data
+     * @type {Set<number>}
+     */
+    filteredNodes;
+
+    constructor(data) {
+        this.data = data;
+        this.idIncrement = 0;
+        this.ids = new Map();
+        this.rawNodesAndLeaves = new Map();
+        this.formattedNodesAndLeaves = new Map();
+        this.filteredNodes = new Set();
+    }
+
+    /**
+     * Create an id to the given node if none have been created previously and return the corresponding id
+     * @param node
+     * @return {number | undefined}
+     */
+    createIdIfNotExist(node) {
+        if (!this.ids.has(node)) {
+            const id = ++this.idIncrement;
+            this.ids.set(node, id);
+            return id;
+        }
+        return this.ids.get(node);
+    }
+
+    /**
+     * Get the hierarchy of a leaf from it first parent
+     * @param node First parent of the leaf
+     * @return {string} Leaf parent list
+     */
+    createHierarchy(node) {
+        const maxRecursion = 10;
+        let recursion = 0;
+
+        const getHierarchy = (parentId) => {
+            recursion++;
+            if (recursion > maxRecursion) {
+                return [];
+            }
+
+            if (!this.formattedNodesAndLeaves.has(parentId)) {
+                return [];
+            }
+
+            const parent = this.formattedNodesAndLeaves.get(parentId);
+            if (parent.parent) {
+                return [...getHierarchy(parent.parent), parent.name];
+            }
+
+            return [parent.name];
+        };
+
+        return getHierarchy(node).join(', ');
+    }
+
+    /**
+     * Transformed routine data into transformable data
+     */
+    buildNodesAndLeaves() {
+        for (let datum of this.data) {
+            const parentNode = datum.source;
+            const node = datum.target;
+            const size = datum.weight;
+            const parentNodeId = this.createIdIfNotExist(parentNode);
+            this.createIdIfNotExist(node);
+
+            if (!this.rawNodesAndLeaves.has(parentNode)) {
+                this.rawNodesAndLeaves.set(parentNode, {});
+            }
+
+            this.rawNodesAndLeaves.set(node, {
+                parent: parentNodeId,
+                size,
+            });
+        }
+    }
+
+    /**
+     * Create a list of node and transformed the raw data into cleanable data
+     */
+    buildFilteredNodes() {
+        for (let rawNodeAndLeaf of this.rawNodesAndLeaves) {
```

Comment:
<!-- Polish comments are like a suggestion, where there is nothing necessarily wrong with the relevant content, there's just some ways to immediately improve the quality.  -->
**polish:** Use destructuring‏